coderd/notifications.go

```diff
@@ -323,6 +327,142 @@ func (api *API) putUserNotificationPreferences(rw http.ResponseWriter, r *http.R
 	httpapi.Write(ctx, rw, http.StatusOK, out)
 }
 
+// @Summary Create user push notification subscription
+// @ID create-user-push-notification-subscription
+// @Security CoderSessionToken
+// @Accept json
+// @Tags Notifications
+// @Param request body codersdk.PushNotificationSubscription true "Push notification subscription"
+// @Param user path string true "User ID, name, or me"
+// @Router /users/{user}/notifications/push/subscription [post]
+// @Success 204
+func (api *API) postUserPushNotificationSubscription(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	user := httpmw.UserParam(r)
+
+	var req codersdk.PushNotificationSubscription
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	notificationJSON, err := json.Marshal(codersdk.PushNotification{
+		Title: "It's working!",
+		Body:  "You've subscribed to push notifications.",
+	})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to marshal notification",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	// Before inserting the subscription into the database, we send a test notification
+	// to ensure the subscription is valid.
+	resp, err := webpush.SendNotificationWithContext(r.Context(), notificationJSON, &webpush.Subscription{
+		Endpoint: req.Endpoint,
+		Keys: webpush.Keys{
+			Auth:   req.AuthKey,
+			P256dh: req.P256DHKey,
+		},
+	}, &webpush.Options{
+		VAPIDPublicKey:  api.PushNotifier.PublicKey(),
+		VAPIDPrivateKey: api.PushNotifier.PrivateKey(),
+	})
```

Comment:
this should be a method inside `push` so we don't need to expose the private key anywhere else